Update to master of upstream, fix some review requests

jmozmoz opened the pull request.

Commit:
Adding support for Throttle configurations optionally in matrix jobs

oleg-nenashev committed:
[FIXED JENKINS-21044] - Throttling blocks the Jenkins queue

Seems the issue was in improper usage of WeakHashMap (see analysis from @centic).
I've managed to reproduce the behavior in the following case:
* There is a big number of jobs/configurations with throttling
* The builds queue is not empty
// Seems that ThrottleQueueTaskDispatcher tries to access the data before the complete loading of the plugin's configuration.

This fix provides an explicit locking of any load operations + manual cleanup of erroneous cache data, which goes to persistence in 1.8.1
Resolves https://issues.jenkins-ci.org/browse/JENKINS-21044

Signed-off-by: Oleg Nenashev <o.v.nenashev@gmail.com>

oleg-nenashev committed:
[JENKINS-21044] - Don't create a new HashMap after the load operation

Signed-off-by: Oleg Nenashev <o.v.nenashev@gmail.com>

abayer committed:
Merge pull request #13 from synopsys-arc-oss/JENKINS_21044_fix

[FIXED JENKINS-21044] - Throttling blocks the Jenkins queue

abayer committed:
[maven-release-plugin] prepare release throttle-concurrents-1.8.2

abayer committed:
[maven-release-plugin] prepare for next development iteration

oleg-nenashev committed:
[JENKINS-13619] - Display default values in multi-config checkboxes

Signed-off-by: Oleg Nenashev <o.v.nenashev@gmail.com>

oleg-nenashev committed:
[JENKINS-13619] - Check configured options for multi-config throttling

Signed-off-by: Oleg Nenashev <o.v.nenashev@gmail.com>

Conflicts:
	src/main/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcher.java

oleg-nenashev committed:
[JENKINS-13619] - Committed missing changes

Signed-off-by: Oleg Nenashev <o.v.nenashev@gmail.com>

Conflicts:
	src/main/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcher.java

oleg-nenashev committed:
Merged configuration options from https://github.com/jenkinsci/throttle-concurrent-builds-plugin/commit/c8280ad3ac3c4faf9b69e0401800813a5953dcfa

Signed-off-by: Oleg Nenashev <o.v.nenashev@gmail.com>

oleg-nenashev committed:
Fixed merge conflicts, prevented "missing data" if someone migrates from a custom build from @papajulio

Signed-off-by: Oleg Nenashev <o.v.nenashev@gmail.com>

oleg-nenashev committed:
Annotate ThrottleMatrixProjectOptions methods

Signed-off-by: Oleg Nenashev <o.v.nenashev@gmail.com>

oleg-nenashev committed:
Fixed tests

Signed-off-by: Oleg Nenashev <o.v.nenashev@gmail.com>

oleg-nenashev committed:
Merge pull request #18 from jenkinsci/throttle-matrix-configurations

[JENKINS-13619] - Throttle matrix configurations

abayer committed:
[maven-release-plugin] prepare release throttle-concurrents-1.8.3

abayer committed:
[maven-release-plugin] prepare for next development iteration

mattmoor committed:
This change fixes an issue introduced with the Matrix support around allowing flyweight tasks to throttle a given category.

Essentially, if a non-Matrix flyweight task in category A was running, we wouldn't count its one-off executor when determining whether another project in category A should be scheduled.

The fix is to count all executors for all tasks that should be throttled instead of only counting one-off executors for Matrix jobs.

cfs-pure committed:
Count builds running on a node's OneOffExecutor list when making
throttling decisions.

oleg-nenashev committed:
Merge pull request #22 from mattmoor/master

This change fixes an issue introduced with the Matrix support around allowing flyweight tasks to throttle a given category

oleg-nenashev committed:
Merge pull request #23 from cfs-pure/master

Include count of  builds running on a node's OneOffExecutor list

oleg-nenashev committed:
Remove the duplicated entry after merging #22 and #23

oleg-nenashev committed:
Added Oleg Nenashev to maintainers

oleg-nenashev committed:
[maven-release-plugin] prepare release throttle-concurrents-1.8.4

oleg-nenashev committed:
[maven-release-plugin] prepare for next development iteration

vitopn committed:
Switch Lists used to store categories to a concurrent-safe implementation in order to avoid ConcurrentModificationExceptions.

stephenc committed:
ignore IDEA directory

oleg-nenashev committed:
Merge pull request #24 from sflyphotobooks/make_categories_concurrent_modification_safe

Switch Lists used to store categories to a concurrent-safe implementatio...

jmozmoz committed:
Merge remote-tracking branch 'upupstream/master'

Conflicts:
	.gitignore
	src/main/java/hudson/plugins/throttleconcurrents/ThrottleJobProperty.java
	src/main/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcher.java
	src/main/resources/hudson/plugins/throttleconcurrents/Messages.properties
	src/test/java/hudson/plugins/throttleconcurrents/ThrottleJobPropertyTest.java

jmozmoz committed:
Implement (most) review requests in
https://github.com/jenkinsci/throttle-concurrent-builds-plugin/pull/9

nstielau commented:
Thanks @jmozmoz!

This is working pretty well for us right now, and don't have a strong driver to update, although the changes look like they make sense.  If there's anything in our branch that you want, it might be easier to get that upstream.  Would love to hear you're use-case though.  Cheers,

-Nick

jmozmoz commented:
My use case is that I have jobs whose workspace directory name is created based on (some) build parameters. Now I would like to make sure that no jobs with identical workspace directory names are running simultaneously. So what I ideally need is the ability to throttle builds based on environment variables (in my case WORKSPACE). But this approach is the best approximation I have found so far.

The next step should probably be a generalization of this approach so that also environment variables can be used as condition for throttling (see these bug reports: https://issues.jenkins-ci.org/browse/JENKINS-25327 https://issues.jenkins-ci.org/browse/JENKINS-23558)

dsavenko committed:
Reduce applicability requirements from AbstractProject to Job & Queue.Task.

dsavenko committed:
Remove needless formatting changes

abayer committed:
Merge pull request #32 from xored/master

Reduce applicability requirements

jmozmoz committed:
Store parameters to compare in a transient list build on save/load

jmozmoz committed:
Changes requested by reviewer:
- Rename limitOneJobByParams to paramsToUseForLimit
- Change style of method header comment

jmozmoz committed:
Merge remote-tracking branch 'upupstream/master'

Conflicts:
	src/main/java/hudson/plugins/throttleconcurrents/ThrottleJobProperty.java
	src/main/java/hudson/plugins/throttleconcurrents/ThrottleQueueTaskDispatcher.java

jmozmoz committed:
Removed obsolet logging.

jmozmoz committed:
Remove compilation error. FIXME: paramsToUseForLimit not loaded into
configuration form

jmozmoz committed:
Remove obsolete code. Add getter for paramsToUseForLimit

jmozmoz committed:
Remove obsolete debugging output.

jmozmoz committed:
Manual merge with upstream master
